agent/checks/check.go

```diff
@@ -707,3 +690,33 @@ func (c *CheckGRPC) Stop() {
 		close(c.stopCh)
 	}
 }
+
+// FailuresHandler keep tracks of successive errors count and ensures
+// that status can be set to critical only once the successive number of failures
+// reaches the given threshold.
+type FailuresHandler struct {
+	FailuresBeforeCritical int
+	Inner                  CheckNotifier
+	Logger                 *log.Logger
+
+	failuresCounter int
+}
+
+func (f *FailuresHandler) updateCheck(checkID types.CheckID, status, output string) {
+
+	if status == api.HealthPassing || status == api.HealthWarning {
+		f.Logger.Printf("[DEBUG] agent: Check %q is %q", checkID, status)
+		f.Inner.UpdateCheck(checkID, status, output)
+		f.failuresCounter = 0
+		return
+	}
+
+	f.failuresCounter++
+	if f.failuresCounter >= f.FailuresBeforeCritical {
+		f.Logger.Printf("[WARN] agent: Check %q is now critical", checkID)
+		f.Inner.UpdateCheck(checkID, status, output)
+		return
+	}
+
+	f.Logger.Printf("[WARN] agent: Check %q is critical but has not reached failure threshold %d/%d", checkID, f.failuresCounter, f.FailuresBeforeCritical)
```

Comment:
A message in the style of "Check %q failed but has not reached failure threshold" would indicate more clearly that the check status is *not* critical.